src/include/storage/sql_table.h

```diff
@@ -27,6 +31,110 @@ class SqlTable {
+     * @return reference to the underlying tuple slot
+     */
+    const TupleSlot &operator*() const { return *current_it_; }
+
+    /**
+     * @return pointer to the underlying tuple slot
+     */
+    const TupleSlot *operator->() const { return &(*current_it_); }
+
+    /**
+     * pre-fix increment.
+     * @return self-reference after the iterator is advanced
+     */
+    SlotIterator &operator++() {
+      TERRIER_ASSERT(!is_end_, "Cannot increment an end iterator");
+      current_it_++;
+      AdvanceOnEndOfDatatable_();
+      return *this;
+    }
+
+    /**
+     * post-fix increment.
+     * @return copy of the iterator equal to this before increment
+     */
+    const SlotIterator operator++(int) {
+      SlotIterator copy = *this;
+      operator++();
+      return copy;
+    }
+
+    /**
+     * Equality check.  Either both iterators are "end" or they must match
+     * on both their observed version and the underlying tuple.
+     * @param other other iterator to compare to
+     * @return if the two iterators point to the same slotcolumn_ids
+     */
+    bool operator==(const SlotIterator &other) const {
+      // First "is_end" check is that both are end, the second protects the iterator check through short-circuit
+      return (is_end_ && other.is_end_) ||
+             (is_end_ == other.is_end_ && txn_version_ == other.txn_version_ && current_it_ == other.current_it_);
+    }
+
+    /**
+     * Inequality check.
+     * @param other other iterator to compare to
+     * @return if the two iterators are not equal
+     */
+    bool operator!=(const SlotIterator &other) const { return !this->operator==(other); }
+
+    DataTable::SlotIterator *GetDataTableSlotIterator() { return &current_it_; }
+
+   private:
+    friend class SqlTable;
+
+    SlotIterator(const common::ConcurrentMap<layout_version_t, DataTableVersion> *tables,
+                 const layout_version_t txn_version, bool is_end)
+        : tables_(tables), current_it_(tables_->Find(txn_version)->second.data_table->begin()) {
+      txn_version_ = txn_version;
+      curr_version_ = txn_version;
+      is_end_ = is_end;
+    }
+
+    /**
+     * Checks if DataTable::SlotIterator is at the end of the table and advances
+     * the iterator to the next table or sets is_end flag as appropriate.  This
+     * refactor is necessary to keep iterator in correct state when used in two
+     * distinct ways:  increment called directly on this object or implicitly
+     * through advancement of the DataTable iterator in SqlTable::Scan.
+     */
+    void AdvanceOnEndOfDatatable_() {
+      if (current_it_ == tables_->Find(curr_version_)->second.data_table->end()) {
+        // layout_version_t is uint32_t so we need to protect against underflow.
+        if (!curr_version_ == 0) {
+          is_end_ = true;
+          return;
+        }
+        curr_version_--;
+        TERRIER_ASSERT(curr_version_ < txn_version_, "Current version must be older than transaction");
```

Comment:
Why put the assertion here? We just decremented the curr_version_. It is unlikely that the current version will become greater or equal to the transaction version unless underflow happens. But you have checked that before.

Reply:
The assertion is here because we had an underflow in the first draft.  I agree that it should probably move to an assertion of `curr_version_ <= txn_version_` at the start of the function.